### ISSUE TITLE
IIS -> Lock .NET Core application DLL file

### ISSUE THREAD

**Issue by radenkozec** [aspnet/IISIntegration#226]
When I try to overwrite .NET Core application DLL file using FTP on production IIS server file is locked and it cannot be overwritten.

To deploy new version I need to stop application in IIS to release lock, then overwrite.
It is not possible to deploy without stopping application.

**Comment by Tratcher** [comment]
See app_offline.htm: https://docs.asp.net/en/latest/hosting/aspnet-core-module.html#asp-net-core-module-app-offline-htm

**Comment by GuardRex** [comment]
If you end up desiring a PowerShell way of doing it, you can use the IIS Admin cmdlets: https://technet.microsoft.com/en-us/library/ee790599.aspx
```
Stop-WebAppPool -Name $appPoolName

... deploy ...

Start-WebAppPool -Name $appPoolName
```

**Comment by radenkozec** [comment]
Thanks for explanation @Tratcher .
@GuardRex I cannot do it because it is shared hosting enviroment and don't have permission to stop AppPool

**Comment by chuchuva** [comment]
Can you make it to 'just work' with msdeploy.exe and Azure? If I understand correctly the web site needs to be restarted to prevent file locking. `-enableRule:AppOffline` works but the whole website is offline for few minutes which is not a great user experience especially given that we deploy few times per day.

See also http://stackoverflow.com/q/40276582/14131

**Comment by Tratcher** [comment]
@chuchuva maybe, but all magic comes with a cost. Previous versions of ASP.NET did some complicated shadow copying to work around the file locking problems.

**Comment by HarelM** [comment]
Magic or not, It worked well. I kinda miss it now after the migration to ASP.NET core...

**Comment by JaCraig** [comment]
Agreed with @HarelM. We've run into issues with this from our automated deployments to just end user experience. We've gone from deploying about 10 times a day with the old MVC to maybe a daily deployment at night and accepting that the people using the Core app are going to be annoyed when we take it offline. While not a show stopper, it's added friction towards Core adoption.

**Comment by Kuqd** [comment]
👍 

Not supporting overlapped recycling is a regression.

**Comment by deadlydog** [comment]
Are there plans to revisit this feature and put it on the road map? It's very inconvenient/unfriendly to force everybody using a .Net Core website to have to manually implement a staging-slot strategy in order to support zero downtime deployments, especially when it's the "happy path" and not even an edge case scenario.

**Comment by ajeckmans** [comment]
We would like to know as well. Putting a app_offline.htm inside the applications directory does not work.

**Comment by DaleMckeown** [comment]
I've only just realised this 'feature' after moving a host of sites ove to AspNetCore. I can't believe it's deemed to be acceptable to take your sites offline for a few minutes every time you want to publish!

This is bad enough for me as a leader of a small team - are those practicing continuous integration on a large scale shunning AspNetCore?

**Comment by davidfowl** [comment]
Are you deploying use FTP or xcopying? Or are you using webdeploy?

**Comment by DaleMckeown** [comment]
I'm publishing via webdeploy to IIS.

**Comment by ajeckmans** [comment]
We are currently working around this issue by just deleting the web.config first (effectively killing the application), but this is not really an acceptable solution in the long run. 

@DaleMckeown ideally with a continuous integration workflow you'd have multiple servers behind a load balancer. You'd then pull one server, update it, put it back en move on to the next. Ofcourse this is not always possible (like in our case), so you'll have to live with a few minutes downtime. In our case the app is back up within 30s, so this is not really an issue.

**Comment by davidfowl** [comment]
@DaleMckeown 

> I'm publishing via webdeploy to IIS.

There are a few options that can be used to make deployment using webdeploy work well (it supports renaming locked files and dropping app offline automatically). Is that the case by default @shirhatti ?

@ajeckmans 

> We are currently working around this issue by just deleting the web.config first (effectively killing the application), but this is not really an acceptable solution in the long run.

Does that mean you're doing xcopy deployment?

**Comment by DaleMckeown** [comment]
> There are a few options that can be used to make deployment using webdeploy work well (it supports renaming locked files and dropping app offline automatically). Is that the case by default @shirhatti ?

Thanks David - sounds better than what I am currently doing (manually stopping the site & app pool in IIS). Can you point to some resources so I can investigate the implications of these approaches?

**Comment by davidfowl** [comment]
@DaleMckeown some info untilI find the source of truth https://github.com/Microsoft/vsts-tasks/issues/5259#issuecomment-346202503

**Comment by ajeckmans** [comment]
@davidfowl We're using webdeploy to deploy to our testserver (which btw never gives us problems), from there we copy the files to our live servers using robocopy

**Comment by guardrex** [comment]
@ajeckmans 

> We are currently working around this issue by just deleting the web.config first

When the *web.config* is removed from the deployment, IIS will serve sensitive files from the deployment. An attacker could simply request files continuously day and night until your 30s window opens up.

@DaleMckeown

Following @ajeckmans advice if you ever want to get into the weeds a bit with a PS approach to do that, you can script the process to drop AppPools one at a time for deployment across a web farm. For an example of how that can be done, see https://github.com/guardrex/aspnetcore-iis-ps-publish ... but just look at that as an experimental example and not a production-quality script. I haven't played with that for a while, but it should (in theory) still work.

**Comment by ajeckmans** [comment]
@guardrex you're right. We copy an app_offline.htm to the dir first, then we remive the web.config, copy over the application, put back the web.config and remove the app_offline (all with a script ofc). Unfortunately just placing the app_offline file in the websites directory does not break the lock on the dlls. We need to remove the web.config, an action we don't need to do for the older asp.net full apps (like old webforms etc)

**Comment by guardrex** [comment]
@ajeckmans That shouldn't work. When the *web.config* file is removed, IIS picks that change up immediately, so it should default requests over to the Static File Module and start serving files. Try it and see if that's what happens ...

1. Add *app_offline.htm*.
2. Confirm that the site serves *app_offline.htm*.
3. Pull the *web.config* out.
4. Request a sensitive file (e.g., `http://localhost:<PORT>/<ASSEMBLY_NAME>.deps.json` ... substituting the PORT and the ASSEMBLY_NAME).
5. You should get the *deps.json* file served if your Static File Module is functioning correctly.

I wouldn't trust merely removing module with ...
```xml
<configuration> 
 <system.webServer> 
   <modules> 
     <remove name="StaticFileModule" /> 
   </modules> 
 </system.webServer> 
</configuration>
```
... since other modules would remain and possibly present other attack vectors. Maybe one could remove all of the non-essential modules, but we're getting into uncharted waters with a move like that just to cover a deployment via removal of *web.config*. **That's never been an option under the official guidance, and so it's completely unsupported.** I recommend PS scripting, for example, to take the AppPool down, or some other strategy.

**Comment by guardrex** [comment]
I guess I should add a note of sympathy to that ... it did raise a few eyebrows from a security perspective. When that deployment layout change was made, it was discussed, including putting the file back into `wwwroot`, see ...

[Discussion for: Publish for IIS changes to web.config location (IISIntegration 158](https://github.com/aspnet/IISIntegration/issues/158)
[Head-check on moving `web.config` back to `wwwroot` (IISIntegration 164)](https://github.com/aspnet/IISIntegration/issues/164)

**Comment by DaleMckeown** [comment]
This leaves me even more confused about what we should be doing in this situation.

What's the current recommendation for publishing AspNetCore sites to IIS via web deploy in a way that can avoid file locks?

**Comment by JasonRodman** [comment]
I would like to know as well. This is becoming a major roadblock to .net core adoption at my organization.

**Comment by shahjay748** [comment]
When my .net core web application is running and I am trying to publish a new build, it gives me error that DLLs are in use. So I have to Stop the app pool, then update the DLLs and start the App Pool.

**Is there any workaround in .Net Core by which I can publish the new Build/DLLs without stopping the App Pool ?**

This is by default supported in .Net framework using shadow copy mechanism. Unfortunately, It's been 2 years since the launch of .Net core and seems like there's still no support for such very basic and required functionality. Is there even any plans to fix this in future ?

Thanks in advance for any help.

**Comment by ajeckmans** [comment]
@guardrex it is indeed scary to break the app to get IIS (or whatever is holding the lock) to release the dll's, however it is even scarier to not be able to push hotfixes to a product. For the time being breaking the app is the only thing we can do, but we're ofc looking into other more modern approaches to handling this :)

@shahjay748 don't hold your breath. If I were te redo the process here, I'd put the app inside a container and just put up a new container and switch traffic and pull the old version down (or something like that). It seems like the more sane thing to do and seeing as .net core is all about new modern ways, just pushing a new version of the app by simply copying over the new files is considered (and I agree) a bit of an arcane way to do it.
Which ofc doesn't help us who are, for the time being, stuck with old processes :)

**Comment by bpetersen1** [comment]
What i do (Not sure if it's a proper way of doing it), is upload the site to another directory and switch paths in iis

**Comment by jkotalik** [comment]
We did some internal investigation on this issue. From what I can tell, ANCM does not hold any files/handles in the deployed application. It seems to be an issue with webdeploy itself, and it flakily fails deployment. I never had it consistently fail to deploy the application after retrying a few times once app_offline was dropped.

**Comment by guardrex** [comment]
Adding a quick PowerShell side note to that: The lock on the app's assembly is always released ... I've never had a problem (*yet!* lol).

@bpetersen1 ... Your "staging" approach has the sweet side effect of offering a quick rollback option if the new deployment flakes out. That should be easy to script if needed.

**Comment by HarelM** [comment]
This is still externally annoying... we have decided to look into a docker solution in order to leave IIS behind.

**Comment by davidfowl** [comment]
I'm going to experiment with a solution. Stay tuned.

**Comment by niico100** [comment]
I'm also having this issue using FTP.

**Comment by davidfowl** [comment]
> I'm also having this issue using FTP.

> What is Microsoft's advice on this?!

Drop an appoffline.htm file, then ftp then remove it.

**Comment by benzhi2011** [comment]
@davidfowl  Yes,that is what i am doing right now with a batch file. It takes few seconds for IIS to release the lock and the copy the files over.

**Comment by benzhi2011** [comment]
@niico100 Basically, the batch file just copy the files to the project directory, and I make a backed folder first before doing that. Before copy the files, put the appoffline.htm to the project folder.
Cause the file locking problem, the copying will make a retry. I am using robcopy
https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/robocopy

**Comment by mscrivo** [comment]
WebDeploy didn't work for me.  I tried using it against my local IIS and it would still complain of locked files.   I then tried using the above powershell commandlets from VS Pre/Post build events, but that didn't work for me either, probably because 32-bit powershell can't modify 64-bit IIS settings?  Anyhow, what does seem to work nicely are the following pre/post build events in my ASP.Net Core project:

Pre-Build Event:
`echo "App Offline" /a > $(ProjectDir)app_offline.htm`

Post-Build Event:
`del $(ProjectDir)app_offline.htm`
